include/fiction/algorithms/simulation/sidb/critical_temperature.hpp

```diff
@@ -0,0 +1,378 @@
+ */
+struct critical_temperature_params
+{
+    /**
+     * Gate based simulation is conducted.
+     */
+    bool gate_based_simulation = true;
+    /**
+     * All physical parameters for physical SiDB simulations.
+     */
+    const sidb_simulation_parameters& sidb_sim_params{};
+    /**
+     * Probability that the ground state is less populated due to temperature. For gate-based simulation, this is the
+     * probability of erroneous calculations of the gate.
+     */
+    double confidence_level{0.99};
+    /**
+     * Simulation stops at max_temperature (room temperature ~300 K).
+     */
+    uint64_t max_temperature{400};
+    /**
+     * Truth table of the given gate (if layout is simulated in `gate-based` mode).
+     */
+    tt truth_table{};
+    /**
+     * Input bit (e.g. 0 -> 00, 1 -> 01, ...).
+     */
+    uint64_t input_bit{};
+};
+
+/**
+ * This struct stores the result of the temperature simulation.
+ *
+ * @paramt Lyt Cell-level layout type.
+ */
+template <typename Lyt>
+struct critical_temperature_stats
+{
+    /**
+     * Critical temperature of the given layout.
+     */
+    double critical_temperature{};
+    /**
+     * Number of physically valid charge configurations.
+     */
+    uint64_t num_valid_lyt{};
+    /**
+     * Energy difference between the ground state and the first (erroneous) excited state.
+     */
+    double energy_between_ground_state_and_first_erroneous = std::numeric_limits<double>::infinity();
+    /**
+     * Prints the simulation results to the given output stream.
+     *
+     * @param out Output stream.
+     */
+    void report(std::ostream& out = std::cout) const
+    {
+        out << fmt::format("critical temperature  = {:.2f} K\n", critical_temperature);
+
+        if (num_valid_lyt != 0)
+        {
+            out << fmt::format("'# of physically valid charge configurations': {} | Energy between ground state and "
+                               "first erroneous: {}\n",
+                               num_valid_lyt, energy_between_ground_state_and_first_erroneous);
+        }
+        else
+        {
+            out << "no state found | if two-state simulation was used, try re-running with three states\n";
+        }
+
+        out << "_____________________________________________________" << std::endl;
+    }
+};
+
+namespace detail
+{
+
+template <typename Lyt>
+class critical_temperature_impl
+{
+  public:
+    critical_temperature_impl(Lyt& lyt, critical_temperature_params& params, critical_temperature_stats<Lyt>& st) :
+            layout{lyt},
+            parameter{params},
+            temperature_stats{st}
+    {
+        static_assert(is_cell_level_layout_v<Lyt>, "Lyt is not a cell-level layout");
+        static_assert(has_sidb_technology_v<Lyt>, "Lyt is not an SiDB layout");
+        static_assert(has_siqad_coord_v<Lyt>, "Lyt is not based on SiQAD coordinates");
+    }
+
+    bool gate_based_simulation()
```

Comment:
This function is still pretty large. Can we break it down into smaller sub-functions? They can either be part of the `detail` namespace or be defined as `private` member functions of this class.

Either way, you could share these functions between the two simulation cases.